docs/architecture/adr-071-bank-v2.md

```diff
@@ -0,0 +1,102 @@
+
+Each denom will be an `x/account` and will implement `OnTransfer`, `OnBurn`, `OnMint` and `OnBalanceOf` hooks. The bank module will check if the denom is an x/account and implement the respective hook. Depending on the hook, th bank module will send a message to the denom account. From there the denom account will return the new balances of the accounts.
+
+We would have the following messages `MsgMint`, `MsgBurn`, `MsgOnTransfer`, `QueryBalance` that would be sent to the denom account. 
+
+How the `x/account` denom implementation would look is as per below:
+
+```go
+func (a Account) RegisterExecuteHandlers(r accountsd.Router) {
+accountsd.RegisterHandler(r, a.OnTransfer)
+accountsd.RegisterHAndler(r. a.Burn)
+}
+
+func (a Account) OnTransfer(ctx context.Context, msg MsgOnTransfer) (MsgOnTransferResponse, error) {
+    ... check if account KYCd...
+}
+```
+
+For example if alice sends `MsgSend{to: bob, amount: 100ATOM}` to bank, bank checks if `ATOM` is an `x/account` implementing the `OnTransfer` hook.
+
+If `ATOM` is an `x/account` implementing the `OnTransfer` hook then bank sends a `MsgTransfer` to the `ATOM` denom account. The `ATOM` denom account returns the new balances of alice and bob.
+
+If `ATOM` is not an `x/account` implementing the `OnTransfer` hook then bank will simply substract amount from alice and add it to bob.
+
+[flow diagram of the above process](flow.png)
+
+
+## Migration Plans
+
+Bank is a widely used module, so getting a v2 needs to be thought thoroughly. In order to not force all dependencies to immediately migrate to bank/v2, the same _upgrading_ path will be taken as for the `gov` module.
+
+This means `cosmossdk.io/bank` will stay one module and there won't be a new `cosmossdk.io/bank/v2` go module. Instead the bank protos will be versioned from `v1beta1` (current bank) to `v2`.
+
+Bank `v1beta1` endpoints will use the new bank v2 implementation for maximum backward compatibility.
+
+The bank `v1beta1` keepers will be deprecated and potentially eventually removed, but its proto and messages definitions will remain.
+
+Additionally, as bank plans to integrate token factory, migrations functions will be provided to migrate from Osmosis token factory implementation (most widely used implementation) to the new bank/v2 token factory.
+
+## Consequences
+
+### Positive
+
+* Simplified interaction with bank APIs
+* Backward comptible changes (no contracts or apis broken)
+* Optional migration (note: bank `v1beta1` won't get any new feature after bank `v2` release)
+
+### Neutral
+
+* Asset implementation not available cross-chain (IBC-ed custom asset should possibly fallback to the default implementation)
+* Many assets may slow down bank balances requests
+
+### Negative
+
+* Temporarily duplicate functionalities as bank `v1beta1` are `v2` are living alongside
+* Difficultity to ever completely remove bank `v1beta1`
```

Comment:
Also: Splintering of fund movement logic.

That is, it sounds like you're expecting each denom to do its own thing when it comes to movement of funds. That makes validation of correctness significantly more difficult, and opens the door to difficult-to-trace bugs that have to be fixed in many places.

Reply:
I'm going to retract this as I now feel that it probably does not need to be included as a negative.